aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,1719 @@
+        // stake pool will inactivate entire pending_inactive stake at `stake::withdraw` to make it withdrawable
+        // however, bypassing the inactivation of excess stake (inactivated but now withdrawn) ensures
+        // the OLC is not advanced indefinitely on `unlock`-`withdraw` paired calls
+        if (can_withdraw_pending_inactive(pool_address)) {
+            // get excess stake before being entirely inactivated
+            let (_, _, _, pending_inactive) = stake::get_stake(pool_address);
+            if (withdrawal_lockup_cycle == pool.observed_lockup_cycle) {
+                // `amount` less excess if withdrawing pending_inactive stake
+                pending_inactive = pending_inactive - amount
+            };
+            // escape excess stake from inactivation
+            stake::reactivate_stake(stake_pool_owner, pending_inactive);
+            stake::withdraw(stake_pool_owner, amount);
+            // restore excess stake to the pending_inactive state
+            stake::unlock(stake_pool_owner, pending_inactive);
+        } else {
+            // no excess stake if `stake::withdraw` does not inactivate at all
+            stake::withdraw(stake_pool_owner, amount);
+        };
+        coin::transfer<AptosCoin>(stake_pool_owner, delegator_address, amount);
+
+        // commit withdrawal of possibly inactive stake to the `total_coins_inactive`
+        // known by the delegation pool in order to not mistake it for slashing
+        let (_, inactive, _, _) = stake::get_stake(pool_address);
+        pool.total_coins_inactive = inactive;
+
+        event::emit_event(
+            &mut pool.withdraw_stake_events,
+            WithdrawStakeEvent {
+                pool_address,
+                delegator_address,
+                amount_withdrawn: amount,
+            },
+        );
+    }
+
+    #[view]
+    /// Return total stake owned by `delegator_address` within delegation pool `pool_address`
+    /// in each of its individual states.
+    public fun get_stake(pool_address: address, delegator_address: address): (u64, u64, u64) acquires DelegationPool {
+        assert_delegation_pool_exists(pool_address);
+        let pool = borrow_global<DelegationPool>(pool_address);
+        let (
+            lockup_cycle_ended,
+            active,
+            pending_inactive,
+            commission_active,
+            commission_pending_inactive
+        ) = calculate_stake_pool_drift(pool);
+
+        active = pool_u64::shares_to_amount_with_total_coins(
+            &pool.active_shares,
+            pool_u64::shares(&pool.active_shares, delegator_address),
+            // exclude operator active rewards not converted to shares yet
+            active - commission_active
+        );
+
+        // if no pending withdrawal, there is neither inactive nor pending_inactive stake
+        let (withdrawal_exists, withdrawal_lockup_cycle) = pending_withdrawal_exists(pool, delegator_address);
```

Comment:
Can we separate this logic for computing inactive and pending_inactive stake a given delegator into a separate function? It'd both improve readability here and can be a useful view function on its own - it shows how much a given delegator has in withdrawal.